app/models/application_form.rb

```diff
@@ -19,7 +20,7 @@ def submitted?
   end
 
   def references_complete?
-    references.select(&:complete?).count == MINIMUM_COMPLETE_REFERENCES
+    references.reload.select(&:complete?).count == MINIMUM_COMPLETE_REFERENCES
```

Comment:
Don't know about "in the app", but several specs fail without it. After some more digging, I think it may be down to how the references are created in those specs - we `create(:application_form, ... reference_count: 0)` and then explicitly `let(:first_reference) { create(:reference, application_form: application_form) }` so that we can set expectations on those instances.

_BUT_ those `first_reference` / `second_reference` instances don't appear in the `references` collection on the `application_form` instance, unless they are explicitly added or the collection is reloaded. Hence expectations on `application_form.references` fail, and by implication, tests that `application_form.references_complete?` also fail.

So I've taken the explicit `reload`s out of the model and into the tests where needed. From a quick scan of the code, I _think_ this will be ok for real runtime behaviour but it would take a lot of time to verify this for sure (unless someone with more direct knowledge of all the paths through the application state changes etc can verify?). Our best proxy for how the app behaves in all circumstances is the tests, and I have them all passing now.... but the only way to be 100% sure is to 🚢 🤡 and try.

If we get any curious behaviour around reference validation in the real world, at least we now have more knowledge of how it can happen, so we're in a good place to deal with it.

Reply:
👍 good idea to move them into the specs! We can tidy those up later.